File: Plugins/AdaptiveCruiseControl/speed.py
from ETS2LA.Utils.Values.numbers import SmoothedValue
from ETS2LA.Utils.settings import Get, Listen
import numpy as np
import logging

# Constants
MU = Get("AdaptiveCruiseControl", "MU", 0.5)  # Coefficient of friction
G = 9.81  # Gravitational acceleration (m/s^2)
MAX_DISTANCE = 150 # Distance at which curvature affects 0%
MIN_DISTANCE = 30 # Distance at which curvature affects 100%

def ListenSettings(settings: dict):
    global MU
    MU = settings.get("MU", 0.5)
    
Listen("AdaptiveCruiseControl", ListenSettings)

def distance_to_point(x1, y1, x2, y2):
    return np.sqrt((x2 - x1)**2 + (y2 - y1)**2)

def calculate_curvature(points, x, z):
    """
    Calculate the curvature for each point in the road segment.
    """
    curvatures = []
    for i in range(1, len(points) - 1):
        # Direction vectors
        v1 = np.array([points[i][0] - points[i - 1][0], points[i][2] - points[i - 1][2]])
        v2 = np.array([points[i + 1][0] - points[i][0], points[i + 1][2] - points[i][2]])
        
        # Angle change
        dot_product = np.dot(v1, v2)
        norm_product = np.linalg.norm(v1) * np.linalg.norm(v2)
        if norm_product == 0:
            continue  # Skip to avoid division by zero
        cos_angle = dot_product / norm_product
        cos_angle = np.clip(cos_angle, -1, 1)  # Ensure valid arccos input
        delta_theta = np.arccos(cos_angle)
        
        # Arc length (average of the two segment lengths)
        delta_s = (np.linalg.norm(v1) + np.linalg.norm(v2)) / 2
        
        # Curvature (1/m)
        if delta_s == 0:
            continue  # Skip to avoid division by zero
        kappa = delta_theta / delta_s
        
        distance = distance_to_point(x, z, points[i][0], points[i][2])
        multiplier = 1 - (distance - MIN_DISTANCE) / (MAX_DISTANCE - MIN_DISTANCE)
        multiplier = np.clip(multiplier, 0, 1)
        
        curvatures.append(kappa * multiplier)
    
    return curvatures

def get_maximum_speed_for_points(points, x, z) -> float:
    """
    Calculate the maximum safe speed based on road curvature.
    """
    
    if len(points) < 3:  # Need at least 3 points to calculate curvature
        return 999
    
    try:
        # Calculate curvatures for all points
        curvatures = calculate_curvature(points, x, z)
        
        if not curvatures:
            return 999
        
        max_curvature = max(curvatures)
        
        # sqrt(MU * G / max_curvature) = max_speed
        try:
            max_speed = np.sqrt(MU * G / max_curvature)  # In m/s
        except:
            return 999

        if max_speed == 0:
            return 999

        return max_speed
    
    except Exception as e:
        logging.exception("Failed to calculate maximum speed")
        return 999

File: Plugins/AdaptiveCruiseControl/Speed.cs
using System.Numerics;
using System.Linq;

namespace AdaptiveCruiseControl;

public static class Speed
{
    private const double G = 9.81; // m/s^2
    private const double MaxDistance = 150; // meters
    private const double MinDistance = 30; // meters

    public static double Mu { get; set; } = 0.5;

    public static double GetMaximumSpeedForPoints(IReadOnlyList<Vector3> points, double x, double z)
    {
        if (points.Count < 3)
            return 999;

        try
        {
            var curvatures = CalculateCurvatures(points, x, z);
            if (curvatures.Count == 0)
                return 999;

            var maxCurvature = curvatures.Max();
            if (maxCurvature <= 0)
                return 999;

            var maxSpeed = Math.Sqrt(Mu * G / maxCurvature);
            if (maxSpeed <= 0)
                return 999;

            return maxSpeed;
        }
        catch
        {
            return 999;
        }
    }

    private static List<double> CalculateCurvatures(IReadOnlyList<Vector3> points, double x, double z)
    {
        var curvatures = new List<double>();
        for (int i = 1; i < points.Count - 1; i++)
        {
            var p0 = points[i - 1];
            var p1 = points[i];
            var p2 = points[i + 1];

            var v1 = new Vector2(p1.X - p0.X, p1.Z - p0.Z);
            var v2 = new Vector2(p2.X - p1.X, p2.Z - p1.Z);

            var dot = Vector2.Dot(v1, v2);
            var norm = v1.Length() * v2.Length();
            if (norm == 0)
                continue;

            var cosAngle = Math.Clamp(dot / norm, -1d, 1d);
            var deltaTheta = Math.Acos(cosAngle);

            var deltaS = (v1.Length() + v2.Length()) / 2d;
            if (deltaS == 0)
                continue;

            var kappa = deltaTheta / deltaS;

            var distance = Math.Sqrt(Math.Pow(x - p1.X, 2) + Math.Pow(z - p1.Z, 2));
            var multiplier = 1 - (distance - MinDistance) / (MaxDistance - MinDistance);
            multiplier = Math.Clamp(multiplier, 0d, 1d);

            curvatures.Add(kappa * multiplier);
        }

        return curvatures;
    }
}
